src/index.ts

```diff
@@ -1,69 +1,215 @@
-    const openURL = (url: string) => {
-      try {
-        window.open(url, '_blank', 'noopener,noreferrer');
-      } catch (error) {
-        console.warn(`Error opening ${url}: ${error}`);
-      }
-    };
-
+  optional: [ILayoutRestorer],
+  activate: (app: JupyterFrontEnd, restorer: ILayoutRestorer | null) => {
     const calculateIcon = (args: IDeployAppArgs) => {
       switch (args.origin) {
         case 'main-menu':
           return undefined;
         case 'context-menu':
+        case 'toolbar':
         case undefined:
         default:
           return deployAppIcon;
       }
     };
 
     app.commands.addCommand(CommandIDs.deployApp, {
-      execute: () => {
+      execute: ({
+        origin,
+        baseUrl = DEFAULT_BASE_URL,
+        queryParameters = {}
+      }: IDeployAppArgs) => {
         const currentWidget = app.shell.currentWidget;
-        const currentNotebookPath =
-          currentWidget && currentWidget instanceof DocumentWidget
-            ? currentWidget.context.path
-            : '';
-        let deployUrl;
-        if (currentNotebookPath !== '') {
-          deployUrl = `/services/japps/create-app?filepath=${encodeURIComponent(currentNotebookPath)}`;
+        const filepath = hasContextPath(currentWidget)
+          ? currentWidget.context.path
+          : '';
+
+        const selectedFilepath =
+          queryParameters.filepath !== undefined
+            ? queryParameters.filepath
+            : filepath;
+
+        const headless = (() => {
+          const value = queryParameters.headless;
+          if (value === undefined) {
+            return 'true';
+          }
+          if (
+            typeof value === 'string' &&
+            ['true', 'false'].includes(value.toLowerCase())
+          ) {
+            return value.toLowerCase();
+          }
+          console.warn(`
+            Invalid 'headless' value: ${value}. Defaulting to 'true'.
+            `);
+          return 'true';
+        })();
+
+        const updatedQueryParameters = {
+          ...queryParameters,
+          headless: headless,
+          filepath: selectedFilepath
+        };
+
+        let mainAreaWidget: MainAreaWidget<DeployAppWidget> | undefined;
+        if (headless === 'true') {
+          try {
+            if (!mainAreaWidget || mainAreaWidget.isDisposed) {
+              const content = new DeployAppWidget({
+                origin: origin,
+                baseUrl: baseUrl,
+                queryParameters: updatedQueryParameters
+              });
+              mainAreaWidget = new MainAreaWidget({ content });
+            }
+
+            if (!tracker.has(mainAreaWidget)) {
+              tracker.add(mainAreaWidget);
+            }
+
+            // non-empty origin implies user has called command e.g. main-menu click.
+            // guards against (what appears to be) command execution from tracker
+            // on session restore.
+            if (origin) {
```

Comment:
This caused me a lot of pain - I'm still trying to get my head around how exactly the tracker/restorer logic works. The codebase has a lot of information, but it's complicated to me. It seems like if I only attach the widget when a user clicks a button (i.e. when `origin` is defined), everything works okay. If I don't include this `origin` statement, the restore logic executes this code many times, resulting in multiple failed GET requests in the console.